arborist/auth.go

```diff
@@ -156,29 +157,70 @@ SELECT coalesce(text2ltree("unnest") <@ allowed, FALSE) FROM (
 		return nil, err
 	}
 
+	if request.ClientId != "" {
+		clientAllowed, err := _authorize(request.stmts, exp, args,
+			`
+SELECT coalesce(text2ltree("unnest") <@ allowed, FALSE) FROM (
+	SELECT array_agg(resource.path) AS allowed FROM client
+	JOIN client_policy ON client_policy.client_id = client.id
+	JOIN policy_resource ON policy_resource.policy_id = client_policy.policy_id
+	JOIN resource ON resource.id = policy_resource.resource_id
+	WHERE client.name = $1
+	AND EXISTS (
+		SELECT 1 FROM policy_role
+		JOIN permission ON permission.role_id = policy_role.role_id
+		WHERE policy_role.policy_id = client_policy.policy_id
+		AND permission.service = $2
+		AND permission.method = $3
+	)
+) _, unnest($4::text[]);
+`,
+			request.ClientId,           // $1
+			request.Service,            // $2
+			request.Method,             // $3
+			pq.Array(resources),        // $4
+		)
+		if err != nil {
+			return nil, err
+		}
+		if !clientAllowed {
+			rv = false
+		}
+	}
+	return &AuthResponse{rv}, nil
+}
+
+func _authorize(stmts *CachedStmts, exp Expression, args []string,
+		query string, qargs ...interface{}) (bool, error) {
+	// run authorization query
+	rows, err := stmts.Query(query, qargs...)
```

Comment:
I think you should be able to just do something like this:
```go
var results []bool
err = db.Select(&results, query)
```
and be able to skip iterating through `rows.Next()`

Reply:
(or I guess probably `pq.BoolArray` or something)

Reply:
Oh here we actually needed a map of `argName: value`, so iteration is hardly avoidable I think. But I added the `Select` to the cached statements anyway ;)